File: 230-kth-smallest-element-in-a-bst.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
var cnt int
var nd *TreeNode

func kthSmallest(root *TreeNode, k int) int {
    cnt = 0
    nd = nil
    tra(root, k)
    return nd.Val
}

func tra(node *TreeNode, k int){
    if node == nil || nd != nil{
        return
    }
    tra(node.Left, k)
    cnt += 1
    if cnt == k{
        nd = node
    }
    tra(node.Right, k)
}

File: bwu/binary_tree/230-kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def kthSmallest(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: int
        """
        stack = []
        node = root
        
        while True:
            if node:
                stack.append(node)
                node = node.left
            else:
                node = stack.pop()
                if k == 1:
                    return node.val
                k -= 1
                node = node.right
                
        return
